code/modules/chemistry/tools/hyposprays.dm

```diff
@@ -59,42 +59,60 @@ var/global/list/chem_whitelist = list("antihol", "charcoal", "epinephrine", "ins
 		if (src.safe && add)
 			check_whitelist(src, src.whitelist)
 		src.UpdateIcon()
+		tgui_process.update_uis(src)
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "Hypospray", "Hypospray")
+			ui.open()
+
+	ui_data(mob/user)
+		. = list()
+		var/datum/reagents/R = src.reagents
+		var/list/reagentData = list(
+			maxVolume = R.maximum_volume,
+			totalVolume = R.total_volume,
+			contents = list(),
+			finalColor = "#000000"
+		)
+
+		var/list/contents = reagentData["contents"]
+		if(istype(R) && R.reagent_list.len>0)
+			reagentData["finalColor"] = R.get_average_rgb()
+			for(var/reagent_id in R.reagent_list)
+				var/datum/reagent/current_reagent = R.reagent_list[reagent_id]
+
+				contents.Add(list(list(
+					name = reagents_cache[reagent_id],
+					id = reagent_id,
+					colorR = current_reagent.fluid_r,
+					colorG = current_reagent.fluid_g,
+					colorB = current_reagent.fluid_b,
+					volume = current_reagent.volume
+				)))
```

Comment:
Not requiring for your PR, just general observation that we should probably have a standard way of serializing for tgui purposes a `var/datum/reagents` rather than every place it's being done having to implement it.

Reply:
I almost addressed this in #7822 with some kind of `tgui_format` proc, wasn't very confident in my own implementation though.